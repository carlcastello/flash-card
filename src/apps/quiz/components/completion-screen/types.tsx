export interface IOwnProps {
  onClose: () => void,
  onReview: () => void,
  classes: {
    
  }
}