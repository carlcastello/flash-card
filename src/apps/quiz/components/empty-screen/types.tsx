export interface IOwnProps {
  onEdit: () => void,
  onClose: () => void,
  classes: {

  }
}