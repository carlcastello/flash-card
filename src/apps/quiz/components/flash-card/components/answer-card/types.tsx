import { FlashcardStatus } from "../../types";

export interface IOwnProps {
  answer: string,
  flashcardStatus: FlashcardStatus,
  ref: any,
  classes: {
    paper: string,
    input: string,
    typography: string
  }
}

export interface IOwnState {
  answer: string,
  hideAnswer: boolean,
}