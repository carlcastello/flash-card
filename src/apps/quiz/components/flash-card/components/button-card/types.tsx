import { FlashcardStatus } from "../../types";

export interface IOwnProps {
  flashcardStatus: FlashcardStatus,
  submit: () => void,
  next: () => void,
  classes: {
    button: string
  }
}

export interface IOwnState {

}