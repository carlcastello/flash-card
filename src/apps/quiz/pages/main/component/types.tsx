import { RouteComponentProps } from 'react-router-dom';

import { FlashcardStatus } from '../../../components/flash-card/types';
import { IQuestionCard } from '../../../../commons/types';

import { QuizStatus } from './enum';

export interface IOwnProps extends RouteComponentProps<any> {
  questions: IQuestionCard[],
  requiredData: string[],

  fetchQuiz: (quizId: string) => void,

  classes: {
    boxContent: string,
    boxContainer: string,
    progressBarContainer: string,
    iconButton: string,
  }
}

export interface IOwnState {
  questionIndex: number,
  progressIndex: number,
  flashcardStatus: FlashcardStatus,
  quizStatus: QuizStatus,
  confirmModalToogle: boolean
}