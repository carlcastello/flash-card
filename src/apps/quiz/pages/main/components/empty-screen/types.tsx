export interface IOwnProps {
  onClose: () => void,
  onAddQuestion: () => void,
  classes: {

  }
}