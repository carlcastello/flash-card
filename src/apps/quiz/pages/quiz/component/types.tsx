import { FlashcardStatus } from "../../../components/flash-card/types";
import { QuizStatus } from "./enum";
import { IFlashcard } from "../../../../commons/types";

export interface IOwnProps {
  flashcards: IFlashcard[],
  classes: {
    boxContent: string,
    boxContainer: string,
    progressBarContainer: string,
    iconButton: string
  }
}

export interface IOwnState {
  questionIndex: number,
  progressIndex: number,
  currentFlashcardStatus: FlashcardStatus,
  quizStatus: QuizStatus
}