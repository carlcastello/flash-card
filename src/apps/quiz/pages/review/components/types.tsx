import { RouteComponentProps } from "react-router-dom";

export interface IOwnProps extends RouteComponentProps<any> {
  classes: {

  }
}