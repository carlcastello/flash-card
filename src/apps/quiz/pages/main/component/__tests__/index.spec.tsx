import React from "react";
import { shallow } from "enzyme";

import { MainQuizPage } from "../index";
import toJson from "enzyme-to-json";
import LoadingScreen from "../../../../../commons/components/loading-screen";
import EmptyScreen from "../../../../components/empty-screen";
import { QuizStatus } from "../enum";
import { FlashcardStatus } from "../../../../components/flash-card/types";


describe("The Quiz Main Page Component", () => {
  
  const quizId: string = "quiz-id";

  let wrapper: any;
  let instance: any;
  let fetchQuizFunction: any;

  let push: any;

  beforeEach(() => {
    fetchQuizFunction = jest.fn();
    push = jest.fn();

    wrapper = shallow(
      <MainQuizPage
        questions={[]}
        requiredData={[]}
        fetchQuiz={fetchQuizFunction}
        classes={{
          boxContent: "box-content",
          boxContainer: "box-container",
          progressBarContainer: "progressBarContainer",
          iconButton: "icon-button",
        }}
        history={{ push } as any}
        location={{} as any}
        match={{
          params: {
            quizId
          }
        } as any}/>
    );
    instance = wrapper.instance();
  });

  describe("the component did mount", () => {
    it("calls fetch quiz function if requiredData and quizId", () => {
      
      wrapper = shallow (
        <MainQuizPage
          questions={[]}
          requiredData={["hello"]}
          fetchQuiz={fetchQuizFunction}
          classes={{
            boxContent: "box-content",
            boxContainer: "box-container",
            progressBarContainer: "progressBarContainer",
            iconButton: "icon-button",
          }}
          history={{ push: jest.fn() } as any}
          location={{} as any}
          match={{
            params: {
              quizId
            }
          } as any}/>
      );


      expect(fetchQuizFunction).toHaveBeenCalledWith(quizId);
    });

    it("doesn't call fetch quiz function if quizId but not requiredData", () => {
      fetchQuizFunction = jest.fn();
      wrapper = shallow (
        <MainQuizPage
          questions={[]}
          requiredData={[]}
          fetchQuiz={fetchQuizFunction}
          classes={{
            boxContent: "box-content",
            boxContainer: "box-container",
            progressBarContainer: "progressBarContainer",
            iconButton: "icon-button",
          }}
          history={{ push: jest.fn() } as any}
          location={{} as any}
          match={{
            params: {
              quizId
            }
          } as any}/>
      );

      expect(fetchQuizFunction).not.toHaveBeenCalled();
    });

    it("doesn't call fetch quiz function if not quizId and not requiredData", () => {
      fetchQuizFunction = jest.fn();
      wrapper = shallow (
        <MainQuizPage
          questions={[]}
          requiredData={[]}
          fetchQuiz={fetchQuizFunction}
          classes={{
            boxContent: "box-content",
            boxContainer: "box-container",
            progressBarContainer: "progressBarContainer",
            iconButton: "icon-button",
          }}
          history={{ push: jest.fn() } as any}
          location={{} as any}
          match={{
            params: {}
          } as any}/>
      );

      expect(fetchQuizFunction).not.toHaveBeenCalled();
    });
    
    it("doesn't call fetch quiz function if requiredData but not quizId", () => {
      fetchQuizFunction = jest.fn();
      wrapper = shallow (
        <MainQuizPage
          questions={[]}
          requiredData={["hello"]}
          fetchQuiz={fetchQuizFunction}
          classes={{
            boxContent: "box-content",
            boxContainer: "box-container",
            progressBarContainer: "progressBarContainer",
            iconButton: "icon-button",
          }}
          history={{ push: jest.fn() } as any}
          location={{} as any}
          match={{
            params: {}
          } as any}/>
      );

      wrapper.update();

      expect(fetchQuizFunction).not.toHaveBeenCalled();
    });
  });

  describe("The render function", () => {
    it("shows a loading screen if requiredData length is not 0", () => {
      wrapper.setProps({ requiredData: ["hello"]});

      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(LoadingScreen)).toHaveLength(1);

      wrapper.setProps({ requiredData: []});

      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(LoadingScreen)).not.toHaveLength(1);
    });

    it("shows an empty screen if question length is 0", () => {
      wrapper.setProps({ questions: []});
      
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(EmptyScreen)).toHaveLength(1);

      wrapper.setProps({ questions: [{
        id: "id",
        question: "question",
        subQuestion: "sub-question",
        hint: "hint",
        answer: "hint",
      }]});

      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(EmptyScreen)).not.toHaveLength(1);
    })

    it("shows the completion screen if quizStatus is complete", () => {
      wrapper.setState({ 
        quizStatus: QuizStatus.COMPLETED,
        questions: [{
          id: "id",
          question: "question",
          subQuestion: "sub-question",
          hint: "hint",
          answer: "hint",
        }]
      });

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe("The Redirects", () => {
    it("calls push on redirectToDashboard", () => {
      instance.redirectToDashboard();
      expect(push).toHaveBeenCalledWith("/");
    });
    
    it("calls push on onEditQuiz", () => {
      instance.onEditQuiz();
      expect(push).toHaveBeenCalledWith(`/dashboard/quiz/${quizId}`);
    });

    it("redirect to dashboard onCloseQuiz", () => {
      wrapper.setState({ quizStatus: QuizStatus.IN_PROGRESS });
      instance.onCloseQuiz();
      expect(push).not.toHaveBeenCalled();

      wrapper.setState({ quizStatus: QuizStatus.COMPLETED });
      instance.onCloseQuiz();
      expect(push).toHaveBeenCalledWith("/");
    });
  });

  describe("The state functions", () => {

    it("toggles confirmModalToogle onCloseIconToggle", () => {
      wrapper.setState({ confirmModalToogle: false });

      instance.onCloseIconToggle();

      expect(wrapper.state().confirmModalToogle).toEqual(true);
    });

    it("updates quizStatus and close the modal onConfirmModal", async () => {
      wrapper.setState({ quizStatus: QuizStatus.IN_PROGRESS }); 

      await instance.onConfirmModal();

      expect(wrapper.state().quizStatus).toEqual(QuizStatus.STOPPED);
      expect(push).toHaveBeenCalledWith("/");

    });

    it("updates currentFlashcardStatus and increment progressIndex onUpdateFlashcard", () => {
      wrapper.setState({
        currentFlashcardStatus: FlashcardStatus.HINT,
        progressIndex: 0,
      });

      instance.onUpdateFlashcard(FlashcardStatus.CORRECT);

      expect(wrapper.state().currentFlashcardStatus).toEqual(FlashcardStatus.CORRECT);
      expect(wrapper.state().progressIndex).toEqual(1);
    });

    it("updats quizStatus to review and reset the state onReviewToggle", () => {
      wrapper.setState({
        quizStatus: QuizStatus.COMPLETED,
        progressIndex: 10,
        questionIndex: 10
      });

      instance.onReviewToggle();
      
      expect(wrapper.state().quizStatus).toEqual(QuizStatus.REVIEW);
      expect(wrapper.state().progressIndex).toEqual(0);
      expect(wrapper.state().questionIndex).toEqual(0);
    });
  });
});