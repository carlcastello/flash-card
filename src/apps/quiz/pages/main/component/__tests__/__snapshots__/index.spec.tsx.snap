// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Quiz Main Page Component The render function shows a loading screen if requiredData length is not 0 1`] = `
<WithStyles(LoadingScreen)>
  Loading Quiz...
</WithStyles(LoadingScreen)>
`;

exports[`The Quiz Main Page Component The render function shows a loading screen if requiredData length is not 0 2`] = `
<WithStyles(EmptyPage)
  onClose={[Function]}
  onEdit={[Function]}
/>
`;

exports[`The Quiz Main Page Component The render function shows an empty screen if question length is 0 1`] = `
<WithStyles(EmptyPage)
  onClose={[Function]}
  onEdit={[Function]}
/>
`;

exports[`The Quiz Main Page Component The render function shows an empty screen if question length is 0 2`] = `
<Styled(MuiBox)>
  <Styled(MuiBox)
    className="box-container"
  >
    <Styled(MuiBox)
      className="box-content"
    >
      <Styled(MuiBox)
        className="progressBarContainer"
        px={5}
      >
        <WithStyles(ProgressBar)
          currentQuestion={0}
          flashcardStatus="DEFAULT"
          totalQuestion={1}
        />
        <WithStyles(ForwardRef(IconButton))
          className="icon-button"
          onClick={[Function]}
        >
          <CloseIcon />
        </WithStyles(ForwardRef(IconButton))>
      </Styled(MuiBox)>
      <WithStyles(Flashcard)
        next={[Function]}
        question={
          Object {
            "answer": "hint",
            "hint": "hint",
            "id": "id",
            "question": "question",
            "subQuestion": "sub-question",
          }
        }
        update={[Function]}
      />
    </Styled(MuiBox)>
  </Styled(MuiBox)>
  <WithStyles(StyledModal)
    isOpen={false}
    onConfirm={[Function]}
    onIgnore={[Function]}
  >
    <Styled(MuiBox)
      pb={4}
    >
      <WithStyles(ForwardRef(Typography))>
        Are you sure you want to close this quiz?
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
  </WithStyles(StyledModal)>
</Styled(MuiBox)>
`;

exports[`The Quiz Main Page Component The render function shows the completion screen if quizStatus is complete 1`] = `
<WithStyles(EmptyPage)
  onClose={[Function]}
  onEdit={[Function]}
/>
`;
